python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,251 @@
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(type_, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
+    return partition_value_as_str.lower() == "true"
+
+
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(type_, partition_value_as_str: str):
+    return int(float(partition_value_as_str))
+
+
+@from_partition_value_to_py.register(FloatType)
+@from_partition_value_to_py.register(DoubleType)
+def _(type_, partition_value_as_str: str):
+    return float(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(StringType)
+def _(type_, partition_value_as_str: str):
+    return partition_value_as_str
+
+
+@from_partition_value_to_py.register(UUIDType)
+def _(type_, partition_value_as_str: str):
+    return uuid.UUID(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(FixedType)
+def _(type_, partition_value_as_str: str):
+    return bytearray(bytes(partition_value_as_str, "UTF-8"))
+
+
+@from_partition_value_to_py.register(BinaryType)
+def _(type_, partition_value_as_str: str):
+    return bytes(partition_value_as_str, "UTF-8")
+
+
+@from_partition_value_to_py.register(DecimalType)
+def _(type_, partition_value_as_str: str):
+    return Decimal(partition_value_as_str)
+
+
+@singledispatch
+def to_bytes(type_, value: Any) -> bytes:
+    """A generic function which converts a built-in python value to bytes
+
+    Args:
+        type_(PrimitiveType): An implementation of the PrimitiveType base class
+        value(Any): The value to convert to bytes
+    """
+    raise TypeError(f"Cannot serialize value, type {type_} not supported: '{value}'")
+
+
+@to_bytes.register(BooleanType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<?", 1 if value else 0)
+
+
+@to_bytes.register(IntegerType)
+@to_bytes.register(DateType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<i", value)
+
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
+@to_bytes.register(TimestamptzType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<q", value)
+
+
+@to_bytes.register(FloatType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<f", value)
+
+
+@to_bytes.register(DoubleType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<d", value)
+
+
+@to_bytes.register(StringType)
+def _(type_, value: Any) -> bytes:
+    return value.encode("UTF-8")
+
+
+@to_bytes.register(UUIDType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(type_, value: Any) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(type_, value: Any) -> bytes:
+    scale = abs(value.as_tuple().exponent)
```

Comment:
does there need to be conversion from the implied scale of value and the scale specified in type?

Reply:
I'm not 100% sure on the requirement here. If a value is provided with a higher scale than the defined type, it feels like just converting it might have implications when lower/upper bound values are deserialized later. I'll need to dig in a bit more here.

Reply:
I think given that fact that we expect this to be used from literals, it seems like the normalize should probably happen at that level, and here there should be an assert that the scales are equal.

Reply:
Added a check and raise for mismatched precision and/or scale.